File: androidApp/src/main/kotlin/zinqq/app/components/Numpad.kt
package zinqq.app.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import zinqq.app.R
import zinqq.app.theme.ZinqqTheme
import zinqq.spike.NumpadKey

/**
 * The PWA's sats-only `Numpad` (U13, KTD-11, R12): a Next CTA above a 3×4
 * digit grid on the elevated dark surface. Key presses feed the shared
 * `numpadDigitReducer` at the call site (8-digit cap, leading-zero collapse
 * live in commonMain, not here — R14-style split of logic vs pixels). All
 * targets are at least 44dp.
 */
@Composable
fun Numpad(
    onKey: (NumpadKey) -> Unit,
    onNext: () -> Unit,
    nextEnabled: Boolean,
    modifier: Modifier = Modifier,
    nextLabel: String = "Next",
) {
    val colors = ZinqqTheme.colors
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp))
            .background(colors.darkElevated)
            .padding(start = 24.dp, end = 24.dp, top = 16.dp, bottom = 24.dp),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(colors.cta)
                .alpha(if (nextEnabled) 1f else 0.3f)
                .clickable(enabled = nextEnabled, onClick = onNext)
                .semantics { contentDescription = nextLabel },
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.Center,
        ) {
            Text(
                text = nextLabel.uppercase(),
                color = colors.onCta,
                fontFamily = ZinqqTheme.fonts.display,
                fontWeight = FontWeight.Bold,
                fontSize = 18.sp,
                letterSpacing = 1.sp,
            )
            Icon(
                painter = painterResource(R.drawable.ic_arrow_right),
                contentDescription = null,
                tint = colors.onCta,
                modifier = Modifier
                    .padding(start = 8.dp)
                    .size(20.dp),
            )
        }

        val rows = listOf(
            listOf("1", "2", "3"),
            listOf("4", "5", "6"),
            listOf("7", "8", "9"),
        )
        Column(
            modifier = Modifier.padding(top = 16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            rows.forEach { row ->
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    row.forEach { digit ->
                        DigitKey(
                            label = digit,
                            onClick = { onKey(NumpadKey.Digit(digit.single())) },
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Box(modifier = Modifier.weight(1f))
                DigitKey(
                    label = "0",
                    onClick = { onKey(NumpadKey.Digit('0')) },
                    modifier = Modifier.weight(1f),
                )
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(64.dp)
                        .clip(RoundedCornerShape(12.dp))
                        .clickable { onKey(NumpadKey.Backspace) }
                        .semantics { contentDescription = "Delete" },
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        painter = painterResource(R.drawable.ic_backspace),
                        contentDescription = null,
                        tint = colors.onDark.copy(alpha = 0.7f),
                        modifier = Modifier.size(28.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun DigitKey(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .height(64.dp)
            .clip(RoundedCornerShape(12.dp))
            .clickable(onClick = onClick)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = label,
            color = ZinqqTheme.colors.onDark,
            fontFamily = ZinqqTheme.fonts.display,
            fontWeight = FontWeight.SemiBold,
            fontSize = 24.sp,
        )
    }
}

File: iosApp/iosApp/Components/Numpad.swift
import Shared
import SwiftUI

// MARK: - Shared reducer adapter

/// A key on the sats numpad, as the views speak it. The shared framework's
/// `NumpadKey` protocol hierarchy stays behind `NumpadReducer` (same adapter
/// discipline as `WalletEvent`), but the reduction itself runs in commonMain
/// — 8-digit cap, leading-zero collapse, and backspace behave identically on
/// every client (U18, R12).
enum NumpadInput {
    case digit(Character)
    case backspace
}

enum NumpadReducer {
    /// Numpad amounts cap at 8 digits (₿99,999,999), matching the PWA.
    static let maxDigits = Int(NumpadReducerKt.NUMPAD_MAX_DIGITS)

    /// Pure reduction over a digit-string state, delegated to the shared
    /// `numpadDigitReducer` (the UI-safe R14 carve-out, like `FormatKt`).
    static func reduce(
        _ prev: String,
        _ input: NumpadInput,
        maxDigits: Int = NumpadReducer.maxDigits
    ) -> String {
        let key: NumpadKey
        switch input {
        case let .digit(char):
            guard let scalar = char.unicodeScalars.first, char.isNumber else { return prev }
            key = NumpadKeyDigit(digit: unichar(scalar.value))
        case .backspace:
            key = NumpadKeyBackspace.shared
        }
        return NumpadReducerKt.numpadDigitReducer(
            prev: prev, key: key, maxDigits: Int32(maxDigits)
        )
    }
}

// MARK: - Component

/// The PWA's sats-only `Numpad` (U18, KTD-11, R12): a Next CTA above a 3×4
/// digit grid on the elevated dark surface. Key presses feed the shared
/// reducer at the call site via `NumpadReducer.reduce` (logic in commonMain,
/// pixels here — the R14-style split). All targets are at least 44pt.
struct Numpad: View {
    let onKey: (NumpadInput) -> Void
    let onNext: () -> Void
    let nextEnabled: Bool
    var nextLabel: String = "Next"

    @Environment(\.zinqqColors) private var colors

    private static let rows: [[Character]] = [
        ["1", "2", "3"],
        ["4", "5", "6"],
        ["7", "8", "9"],
    ]

    var body: some View {
        VStack(spacing: 16) {
            nextButton
            VStack(spacing: 8) {
                ForEach(Self.rows, id: \.self) { row in
                    HStack(spacing: 8) {
                        ForEach(row, id: \.self) { digit in
                            digitKey(digit)
                        }
                    }
                }
                HStack(spacing: 8) {
                    Color.clear
                        .frame(maxWidth: .infinity)
                        .frame(height: 64)
                    digitKey("0")
                    backspaceKey
                }
            }
        }
        .padding(.init(top: 16, leading: 24, bottom: 24, trailing: 24))
        .frame(maxWidth: .infinity)
        .background(colors.darkElevated)
        .clipShape(TopRoundedShape(radius: 16))
    }

    private var nextButton: some View {
        Button(action: onNext) {
            HStack(spacing: 8) {
                Text(nextLabel.uppercased())
                    .font(ZinqqFont.display(18, weight: .bold))
                    .kerning(1)
                Image(systemName: "arrow.right")
                    .font(.system(size: 18, weight: .bold))
            }
            .foregroundColor(colors.onCta)
            .frame(maxWidth: .infinity)
            .frame(height: 56)
            .background(colors.cta)
            .clipShape(RoundedRectangle(cornerRadius: 12))
            .opacity(nextEnabled ? 1 : 0.3)
        }
        .disabled(!nextEnabled)
        .accessibilityLabel(nextLabel)
    }

    private func digitKey(_ digit: Character) -> some View {
        Button {
            onKey(.digit(digit))
        } label: {
            Text(String(digit))
                .font(ZinqqFont.display(24, weight: .semibold))
                .foregroundColor(colors.onDark)
                .frame(maxWidth: .infinity)
                .frame(height: 64)
                .contentShape(RoundedRectangle(cornerRadius: 12))
        }
        .accessibilityLabel(String(digit))
    }

    private var backspaceKey: some View {
        Button {
            onKey(.backspace)
        } label: {
            Image(systemName: "delete.backward")
                .font(.system(size: 24))
                .foregroundColor(colors.onDark.opacity(0.7))
                .frame(maxWidth: .infinity)
                .frame(height: 64)
                .contentShape(RoundedRectangle(cornerRadius: 12))
        }
        .accessibilityLabel("Delete")
    }
}
